tutor/commands/upgrade/compose.py

```diff
@@ -166,16 +167,53 @@ def upgrade_from_olive(context: click.Context, config: Config) -> None:
 
     click.echo(fmt.title(f"Upgrading MySQL to {intermediate_mysql_docker_image}"))
 
-    # Revert the MySQL image to build the data dictionary on v8.1
-    mysql_docker_image = config["DOCKER_IMAGE_MYSQL"]
-    config["DOCKER_IMAGE_MYSQL"] = intermediate_mysql_docker_image
+    # We start up a mysql-8.1 container to build data dictionary to preserve
+    # the upgrade order of 5.7 -> 8.1 -> 8.4
+    # Use the mysql-8.1 context so that we can clear these filters later on
+    with hooks.Contexts.app("mysql-8.1").enter():
+        hooks.Filters.ENV_PATCHES.add_item(
+            (
+                "local-docker-compose-services",
+                """
+mysql-8.1:
+  extends: mysql
+  image: docker.io/mysql:8.1.0
+  command: >
+    mysqld
+    --character-set-server=utf8mb3
+    --collation-server=utf8mb3_general_ci
```

Comment:
curious: What happens here if someone has already changed the connection strings on their Mysql57?

Reply:
I am not sure as I have not tested this out, however, we would want this container to be as similar to the one already being launched in tutor v17. I'm not sure we want to cater to every modification a user might have made.

Reply:
Right.